### PR TITLE
Streamline Testing with `pytest-cov` and `pytest` Defaults

### PR DESCRIPTION
**Pull request by WaelKarkoub**
## Why are these changes needed?

This PR aims to simplify the testing process and reduce redundancy in the GitHub workflows.

This PR does the following:
- Migrates to `pytest-cov`: Instead of relying on the coverage package, this PR adds `pytest-cov` plugin. `pytest-cov` works well with `pytest-xdist`, allowing us to run tests concurrently if the CI environment supports it.
- Unifies test execution: The workflows have been updated to use the `pytest` command directly, replacing the previous `coverage -a -m` invocation. This unifies the testing process and simplifies the workflows.

## Update 1:
Copy pasting from discord: 

There's seems to be a silent failure in master:
- with coverage the github action step passes even though the tests are failing.
- with pytest-cov, the github action step fails correctly

The environment variable `AUTOGEN_USE_DOCKER` was not being used properly in these tests, so I added some fixes:
- added  a fix in `test_commandline_code_executor.py`,
- added a fix in `test/test_code_utils.py`

## Update 2:
- There is a bug in Disc cache that doesn't handle windows path correctly, so I added a fix there as well.
- This also led to a CI a bug `PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmpcq4z30pm\\42\\cache.db'`, for windows-2019 and python 3.12 for image gen cache test. I added a skip for that condition till I figure out the env. 

## Update 3:
- Dropped openai version to `openai>=1.3`
- Mypy ignores `openai_utils`.

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Review by ekzhu — APPROVED** [review]

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -39,9 +39,9 @@ jobs:
       - name: Install packages and dependencies
         run: |
           python -m pip install --upgrade pip wheel
-          pip install -e .
+          pip install -e .[test]
           python -c "import autogen"
-          pip install pytest mock
+          pip install mock
```
Comment (sonichi) [inline .github/workflows/build.yml:44]: how about keeping line 42 and modifying line 44 into
`pip install pytest-cov mock`?
Only install required packages.

File: .github/workflows/contrib-openai.yml [hunk 2]
```diff
@@ -38,9 +39,8 @@ jobs:
         run: |
           docker --version
           python -m pip install --upgrade pip wheel
-          pip install -e .
+          pip install -e .[test]
```
Comment (sonichi) [inline .github/workflows/contrib-openai.yml:42]: similar

File: .github/workflows/contrib-openai.yml [hunk 3]
```diff
@@ -85,9 +84,8 @@ jobs:
...
```
Comment (sonichi) [inline .github/workflows/contrib-openai.yml:87]: similar

File: .github/workflows/contrib-openai.yml [hunk 4]
```diff
@@ -167,18 +162,16 @@ jobs:
         run: |
           docker --version
           python -m pip install --upgrade pip wheel
-          pip install -e .[teachable]
+          pip install -e .[test,teachable]
```
Comment (sonichi) [inline .github/workflows/contrib-openai.yml:165]: similar

File: .github/workflows/contrib-openai.yml [hunk 5]
```diff
@@ -205,9 +198,8 @@ jobs:
         run: |
           docker --version
           python -m pip install --upgrade pip wheel
-          pip install -e .
+          pip install -e .[test]
```
Comment (sonichi) [inline .github/workflows/contrib-openai.yml:201]: similar

File: .github/workflows/contrib-openai.yml [hunk 6]
```diff
@@ -246,9 +237,8 @@ jobs:
         run: |
           docker --version
           python -m pip install --upgrade pip wheel
-          pip install -e .[websurfer]
+          pip install -e .[test, websurfer]
```
Comment (sonichi) [inline .github/workflows/contrib-openai.yml:240]: similar

File: .github/workflows/contrib-openai.yml [hunk 7]
```diff
@@ -257,93 +247,88 @@ jobs:
...
+          ref: ${{ github.event.pull_request.head.sha }}
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Install packages and dependencies
+        run: |
+          docker --version
+          python -m pip install --upgrade pip wheel
+          pip install -e .[test]
```
Comment (sonichi) [inline .github/workflows/contrib-openai.yml:277]: similar

File: .github/workflows/contrib-openai.yml [hunk 8]
```diff
@@ -360,18 +345,16 @@ jobs:
         run: |
           docker --version
           python -m pip install --upgrade pip wheel
-          pip install -e .
+          pip install -e .[test]
```
Comment (sonichi) [inline .github/workflows/contrib-openai.yml:348]: similar

File: .github/workflows/contrib-tests.yml [hunk 9]
```diff
@@ -38,7 +38,7 @@ jobs:
       - name: Install packages and dependencies for all tests
         run: |
           python -m pip install --upgrade pip wheel
-          pip install pytest
+          pip install .[test]
```
Comment (sonichi) [inline .github/workflows/contrib-tests.yml:41]: similar

File: .github/workflows/openai.yml [hunk 10]
```diff
@@ -47,9 +48,8 @@ jobs:
         run: |
           docker --version
           python -m pip install --upgrade pip wheel
-          pip install -e.
```
Comment (sonichi) [inline .github/workflows/openai.yml:50]: restore this line

Comment (WaelKarkoub) [reply .github/workflows/openai.yml:50]: Good catch, thank you!

File: .github/workflows/build.yml [hunk 11]
```diff
@@ -71,8 +71,7 @@ jobs:
         if: matrix.python-version == '3.10'
         run: |
           pip install -e .[test,redis,websockets]
-          coverage run -a -m pytest test --ignore=test/agentchat/contrib --skip-openai --durations=10 --durations-min=1.0
-          coverage xml
+          pytest test --ignore=test/agentchat/contrib --skip-openai --durations=10 --durations-min=1.0
```
Comment (ekzhu) [inline .github/workflows/build.yml:74]: Can you switch back to coverage and see if this is the issue?

**Comment by WaelKarkoub** [comment]
@ekzhu @sonichi added details in the description of the PR. TLDR; there was a silent docker+windows bug in the CI, `AUTOGEN_USE_DOCKER` wasn't being used properly by the tests. This PR currently only fails the type check CI, but that will get fixed when https://github.com/microsoft/autogen/pull/2455 gets merged in.

**Comment by sonichi** [comment]
> @ekzhu @sonichi added details in the description of the PR. TLDR; there was a silent docker+windows bug in the CI, `AUTOGEN_USE_DOCKER` wasn't being used properly by the tests. This PR currently only fails the type check CI, but that will get fixed when #2455 gets merged in.

Shall we merge #2455 now and fix the remaining issue in this PR? Including the openai version?

**Comment by WaelKarkoub** [comment]
> > @ekzhu @sonichi added details in the description of the PR. TLDR; there was a silent docker+windows bug in the CI, `AUTOGEN_USE_DOCKER` wasn't being used properly by the tests. This PR currently only fails the type check CI, but that will get fixed when #2455 gets merged in.
> 
> Shall we merge #2455 now and fix the remaining issue in this PR? Including the openai version?

@sonichi I believe so. I will make changes to my PR once https://github.com/microsoft/autogen/pull/2455 goes in, as that PR made few changes to `.github/workflows/contrib-openai.yml` and `.github/workflows/contrib-tests.yml`

**Comment by sonichi** [comment]
> > > @ekzhu @sonichi added details in the description of the PR. TLDR; there was a silent docker+windows bug in the CI, `AUTOGEN_USE_DOCKER` wasn't being used properly by the tests. This PR currently only fails the type check CI, but that will get fixed when #2455 gets merged in.
> > 
> > 
> > Shall we merge #2455 now and fix the remaining issue in this PR? Including the openai version?
> 
> @sonichi I believe so. I will make changes to my PR once #2455 goes in, as that PR made few changes to `.github/workflows/contrib-openai.yml` and `.github/workflows/contrib-tests.yml`

Thanks. Merging #2455

**Comment by sonichi** [comment]
@WaelKarkoub to be clear, we want to keep supporting old openai versions and ignore the type check for now.

File: .github/workflows/contrib-openai.yml [hunk 12]
```diff
@@ -67,8 +68,7 @@ jobs:
           AZURE_OPENAI_API_BASE: ${{ secrets.AZURE_OPENAI_API_BASE }}
           OAI_CONFIG_LIST: ${{ secrets.OAI_CONFIG_LIST }}
         run: |
-          coverage run -a -m pytest -k test_retrievechat test/agentchat/contrib/retrievechat
-          coverage xml
+          pytest -k test/agentchat/contrib/retrievechat/ test/agentchat/contrib/retrievechat
```
Comment (sonichi) [inline .github/workflows/contrib-openai.yml:71]: This line doesn't look right.

**Review by sonichi — COMMENTED** [review]
Thanks. I have one minor question.

**Comment by WaelKarkoub** [comment]
@sonichi @ekzhu errors are fixed. I uncovered a couple of other bugs that I added fixes to (e.g. disc cache file path handling doesn't work with Windows). Just need someone to sanity check one more time and good to merge

**Comment by WaelKarkoub** [comment]
@sonichi we're almost there! `test/oai/test_client_stream.py::test_chat_tools_stream` is a flaky test (even for me), and it doesn't test what it's supposed to be testing. I added a fix for it as well

**Comment by sonichi** [comment]
@WaelKarkoub Thanks. There is some notebook test failure which I will fix.
@ekzhu @thinkall this is one reason we don't want to use a single config in the notebook.

**Review by sonichi — APPROVED** [review]

**Pull request merged**